### ISSUE TITLE
incorrect dot grammar: nodes not identified in graph

### ISSUE THREAD

**Issue by GoogleCodeExporter** [pydot/pydot#70]
What steps will reproduce the problem?
1. Create this dot file:

``` dot
graph condiments {
    ketchup -- mayonnaise -- tartar -- lemon
    ketchup -- bbq
}
```
1. Create and run this python script:

``` python
import sys
import pydot

# first arg is the dot file
if len(sys.argv) < 2:
    exit()

g = pydot.graph_from_dot_file(sys.argv[1])

print g.get_node_list()
```

What is the expected output? What do you see instead?

The output is an empty list, I expected a list of nodes (`ketchup`, `bbq`, etc.)

What version of the product are you using? On what operating system?

pydot-1.0.28 with Python 2.7.1, OS X

Please provide any additional information below.

Original issue reported on code.google.com by `olliegla...@gmail.com` on 9 Apr 2012 at 6:11

**Comment by johnyf** [comment]
Miscellaneous comment: may be useful to replace `pyparsing` with `ply` as the parser generator. There already exist `ply` parsers for DOT, for example [in `grandalf`](https://github.com/bdcht/grandalf/blob/900b72d59c202daaaea8fdc43f2577efafc6b5d8/grandalf/utils/dot.py).

**Comment by johnyf** [comment]
This is not a matter of preference, but about the correctness of the grammar and parser generator (unfortunately `pyparsing` for now, until it is dropped, in favor of `ply`). This is an error with respect to the [grammar](http://www.graphviz.org/doc/info/lang.html) of the DOT language.

As of `pydot==1.2.3.dev0-3a8ee3d414fbe447af59cce753e71bf2fe050771`, the following minimal example demonstrates the grammar error:

```dot
graph condiments {
    a -> b
}
```

then:

```python
import pydot

g = graph_from_dot_file('unidentified.dot')[0]
g.to_string()
```

prints:

```
'graph condiments {\n}\n'
```

as opposed to the same for the dot file:

```dot
graph condiments {
    a
}
```

A little bit more generally: `pydot.dot_parser` does **not** parse any edges at all.

**Comment by peternowee** [comment]
The issue originally reported here is that `get_node_list()` does not return implicitly defined nodes. This is still the case with current pydot 1.4.1. This issue seems to be a duplicate of pydot/pydot#214 and pydot/pydot#13. The `get_nodes()` method mentioned in pydot/pydot#214 is actually just a call to the `get_node_list()` method mentioned here. I am closing this issue in favor of the more recent pydot/pydot#214.

@johnyf I see you made comments about changing the parser in several issues. Sounds interesting. Maybe you can open a new issue/PR for that if you still feel that would be a good improvement. Btw, I [cannot reproduce the example](https://github.com/pydot/pydot/issues/70#issuecomment-231661094) you give above of edge `a -> b` not being printed by `to_string()`, neither in pydot 1.4.1+pyparsing 2.4.7, nor in combinations of pydot 1.2.2 (released 20160701) or 1.2.3.dev0-3a8ee3d (20160710, used for your comment above) with pyparsing 2.0.2 (20140413) or 2.1.5 (20160613), all on  Python 3.7.3.

**Comment by johnyf** [comment]
Indeed, I cannot reproduce my example either, using `pydot` at commit 3a8ee3d414fbe447af59cce753e71bf2fe050771 and `pyparsing == 2.1.4` (the minimum from `install_requires` in `setup.py` in that commit), on Python 3.9.0+. Instead, for the `dot` file contents:

```dot
graph condiments {
    a -> b
}
```

and the Python code:

```python
import pydot
g = pydot.graph_from_dot_file('unidentified.dot')[0]
print(g.to_string())
```

I get the output:

```python
'graph condiments {\na -- b;\n}\n'
```

and when I change the `dot` file contents to:

```dot
digraph condiments {
    a -> b
}
```

I get the output from the Python code:

```python
'digraph condiments {\na -> b;\n}\n'
```
